### DISCUSSION TITLE
How to visualize mappings done by the Mapstruct

### DISCUSSION THREAD

**Discussion by KallenRoskat** [mapstruct/mapstruct#3740] [Q&A] [answered]
Hello,

I have been searching solution for how to visualize mappings done by the Mapstruct without success. I wonder if someone already had similar problem and has solved it? By the visualization I mean if I can find list where the source property is found in the newly mapped object.

Of course I can develop a script for the source code, collect stuff from annotations or implementation files, collect mappings and relations to technical format that can be visualized by the visualization tools. But I wonder if there is a such tool/script already. Developing one for large and complicated mappings would not be a nice task :).

**Comment by zyberzebra** [comment]
Can you provide an example of what you mean by visualization?

**Comment by KallenRoskat** [comment]
Hello,

When Mapstruct is doing mappings it will generate mappings "log" file in some visualization tool format like GraphViz (DOT Language).

"log file": Collected mappings as .js file.
`
--- Mermaid.js example.
classDiagram
namespace Accounts {
    class Bank
    class Internal
    class BankId
}
namespace IOAccounts {
class InternationAccount
class LocalAccount
}

namespace Banks {
    class BankList
}

%% Object mappings
Bank --> InternationAccount
Internal --> LocalAccount
BankId --> BankList
BankId --> LocalAccount
`

Visualized mappings:
![Mapping](https://github.com/user-attachments/assets/9313018d-2c14-43f8-822a-481b0049e958)

**Comment by filiphr** [comment, marked as answer]
In theory you could achieve this by implementing a custom `org.mapstruct.ap.internal.processor.ModelElementProcessor` (keep in mind that this is not public API) and generate some visualization if you want to. You'll need to hook into around that same time as the `org.mapstruct.ap.internal.processor.MapperRenderingProcessor`

**Reply by KallenRoskat** [reply]
Thank you. Will look that.

**Comment by dotxyteam** [comment]
It doesn't really answer your question, but you can also use this visual mapping tool, which can coexist with MapStruct in your code:
https://github.com/dotxyteam/JVisualMapper
